apps/epl/src/epl_timeline.erl

```diff
@@ -0,0 +1,109 @@
+%%% Copyright (c) 2017, erlang.pl
+%%%-------------------------------------------------------------------
+%%% @doc
+%%% Tracking all timeline observers
+%%% @end
+%%%-------------------------------------------------------------------
+-module(epl_timeline).
+-behaviour(gen_server).
+
+%% API
+-export([start_link/0,
+         subscribe/0,
+         unsubscribe/0,
+         handle_pid/2]).
+
+%% gen_server callbacks
+-export([init/1,
+         handle_call/3,
+         handle_cast/2,
+         handle_info/2,
+         terminate/2,
+         code_change/3]).
+
+-record(state, {subscribers = [],
+                timelines = []}).
+
+%%%===================================================================
+%%% API functions
+%%%===================================================================
+
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+subscribe() ->
+    gen_server:cast(?MODULE, {subscribe, self()}).
+
+unsubscribe() ->
+    gen_server:cast(?MODULE, {unsubscribe, self()}).
+
+handle_pid(Action, P) when is_list(P) ->
+    try list_to_pid(P) of
+        Pid -> gen_server:cast(?MODULE, {Action, Pid})
+    catch
+        error:badarg -> noop
+    end;
+
+handle_pid(Action, Pid) when is_binary(Pid) ->
+    handle_pid(Action, binary_to_list(Pid)).
+
+%%%===================================================================
+%%% gen_server callbacks
+%%%===================================================================
+init([]) ->
+    ok = epl:subscribe(),
+    {ok, #state{}}.
+
+handle_call(_Request, _From, State) ->
+    {reply, ok, State}.
+
+handle_cast({subscribe, Pid}, State = #state{subscribers = Subs}) ->
+    {noreply, State#state{subscribers = [Pid|Subs]}};
+handle_cast({unsubscribe, Pid}, State = #state{subscribers = Subs}) ->
+    NextSubs = lists:delete(Pid, Subs),
+    {noreply, State#state{subscribers = NextSubs}};
+handle_cast({add, Pid}, State = #state{timelines = Timelines}) when is_pid(Pid) ->
+    T = case lists:any(fun({P, _}) -> P == Pid end, Timelines) of
+        true -> Timelines;
+        _    ->
+                {ok, Timeline} = epl_timeline_observer:start_link(Pid),
```

Comment:
I'm wondering if we really need `epl_timeline_observer` processes at all. As I see, both `epl_timeline` and `epl_timeline_observer`s are triggered by events sent by `epl_tracer`. When `epl_timeline` receives such event, it sequentially asks each `epl_timeline_observer` for history of a process which it is responsible for. But then each of `epl_timeline_observers` looks for PID it is responsible for in aggregated timeline results, when we are interested in all results anyway, which seems redundant (filtering results to aggregate them again).

The only performance gain I see is the parallelisation of formatting states and messages into binaries, but this could be tackled easier by spawning multiple processes to do formatting in parallel and than aggregate results using simple `spawn_link` in relevant place (in Elixir we would use `Task` module).

By removing dangling `epl_timeline_observer` processes we would be freed from the burden of debugging N processes instead of one, if such necessity occurs. Another advantage would slight improvement in memory consumption - processes are cheap but not free 😄 

What do you think about it? It's just a suggestion, I might have missed some points.